crates/metal-lexer/src/token.rs

```diff
@@ -0,0 +1,264 @@
+
+#[rustfmt::skip]
+#[derive(Logos, Debug, PartialEq)]
+#[logos(error = Error)]
+#[logos(skip " \t\r\n")]
+#[logos(subpattern int = r"[0-9](_?[0-9])*_?")]
+#[logos(subpattern float = r"(?&int)\.(?&int)")]
+#[logos(subpattern string = "(\"[^\"\r\n]*\")|('[^'\r\n]*')")]
+pub enum TokenKind<'src> {
+    /* Literals */
+
+    #[regex(r"(\p{XID_Start}|_)\p{XID_Continue}*")]
+    Ident(&'src str),
+
+    #[regex("(?&int)", convert)]
+    NumLit(u64),
+
+    #[regex("0b[01](_?[01])*", convert_radix::<2>)]
+    BinaryNumLit(u64),
+
+    #[regex("0x[0-9abcdefABCDEF](_?[0-9abcdefABCDEF])*", convert_radix::<16>)]
+    HexNumLit(u64),
+
+    #[regex(r"(?&float)", convert)]
+    FloatLit(f64),
+
+    #[regex(r"(?&float)[eE][+-]?(?&int)", convert)]
+    EFloatLit(f64),
+
+    #[regex("(?&string)", trim_quotes::<false>)]
+    StrLit(&'src str),
+
+    #[regex("true|false", convert)]
+    BoolLit(bool),
+
+    /* Comparison */
+
+    #[token("<")]
+    LessThan,
+
+    #[token("<=")]
+    LessThanOrEqual,
+
+    #[token(">")]
+    GreaterThan,
+
+    #[token(">=")]
+    GreaterThanOrEqual,
+
+    #[token("==")]
+    Equal,
+
+    #[token("!=")]
+    NotEqual,
+
+    /* Logical */
+
+    #[token("&&")]
+    And,
+
+    #[token("||")]
+    Or,
+
+    #[token("!")]
+    Not,
+
+    /* Bitwise operators */
+
+    #[token("&")]
+    BitwiseAnd,
+
+    #[token("|")]
+    BitwiseOr,
+
+    #[token("~")]
+    BitwiseNot,
+
+    #[token("^")]
+    BitwiseXor,
+
+    #[token("<<")]
+    BitwiseShiftLeft,
+
+    #[token(">>")]
+    BitwiseShiftRight,
+
+    /* Math */
+
+    #[token("+")]
+    Plus,
+
+    #[token("-")]
+    Minus,
+
+    #[token("*")]
+    Multiply,
+
+    #[token("**")]
+    Power,
+
+    #[token("/")]
+    Divide,
+
+    #[token("%")]
+    Remainder,
+
+    /* Assignment */
+
+    #[token("=")]
+    Assignment,
+
+    #[token("+=")]
+    PlusAssignment,
+
+    #[token("-=")]
+    MinusAssignment,
+
+    #[token("*=")]
+    MultiplyAssignment,
+
+    #[token("**=")]
+    PowerAssignment,
+
+    #[token("/=")]
+    DivideAssignment,
+
+    #[token("%=")]
+    RemainderAssignment,
+
+    #[token("&=")]
+    BitwiseAndAssignment,
+
+    #[token("|=")]
+    BitwiseOrAssignment,
+
+    #[token("~=")]
+    BitwiseNotAssignment,
+
+    #[token("^=")]
+    BitwiseXorAssignment,
+
+    #[token("<<=")]
+    BitwiseShiftLeftAssignment,
+
+    #[token(">>=")]
+    BitwiseShiftRightAssignment,
+
+    /* Control flow */
+
+    #[token("if")]
+    If,
+
+    #[token("else")]
+    Else,
+
+    #[token("for")]
+    For,
+
+    #[token("while")]
+    While,
+
+    #[token("break")]
+    Break,
+
+    #[token("continue")]
+    Continue,
+
+    #[token("return")]
+    Return,
+
+    #[token("switch")]
+    Switch,
+
+    /* Keywords */
+
+    #[token("const")]
+    Const,
+
+    #[token("def")]
+    Def,
+
+    #[token("defer")]
+    Defer,
+
+    #[token("enum")]
+    Enum,
+
+    #[token("func")]
+    Func,
```

Comment:
doesn't `def` already define a function though? Unless this does something different?

Reply:
I decided to support both while we prototype. Can remove it if you want

Reply:
> I decided to support both while we prototype. Can remove it if you want

Depends on you, I think it's better to just keep one to avoid any future issues in user code.